File: packages/shared/src/content/schema/config/adminFriend.ts
import { z } from "zod";

/**
 * ⭐【`config.lobby-rally@1`】—— **大廳集合令**（GH#492）。
 *
 * owner 2026-08-21 逐字：
 *
 * > 「**創建房間最重要的就是拉人進來**，請你將**所有線上在大廳的人都跳出確認視窗**
 * >  是否進入房間一起開始，同意後就一起進入開始遊戲，**最多等 10 秒**，
 * >  **包含 vs bot**，若有其他玩家一起進入房間遊戲，也請出現**明顯提示姓名與積分、
 * >  所選英雄**，**每回合結算也都要特別再提示一次**，因為**有可能斷線離開或連線
 * >  回來房間繼續遊戲**」
 *
 * ⭐ **2026-08-21 owner 反轉了語意**（逐字）：
 *
 * > 「你說的是對的，**預設是加入，五秒是讓人按否定的**」
 *
 * ⇒ 倒數結束 = **加入**，⛔ 不是「放棄」。視窗的主要按鈕是「**不要**」，
 * 沒有互動就進房。⭐ 這是 owner **獨立的**裁決（「預設是加入」），
 * ⛔ 與窗口長短無關 —— 窗口 2026-08-21 晚上已從 5 秒改回 **10 秒**，而 opt-out 不變，
 * 而 owner 的原話是「**創建房間最重要的就是拉人進來**」。
 * `joinMode` 是那個反轉的一鍵 rollback（`opt-in` = 2026-08-21 早上那一版）。
 *
 * ⭐ 「**最多等 10 秒**」是這一族唯一被 owner 說死的數字（後來他自己改成 5），
 * 其餘每一格都是**決策點**（第一守則：「心裡出現要選 A 還是 B」的那些），
 * 所以它們是欄位而不是常數 —— 而且 `enabled` 那一格就是第〇·六守則要的
 * **一鍵 rollback**：關掉它，一鍵開打回到 2026-08-21 之前那條「立刻開，不等人」的路。
 *
 * ⚠️ 這一份**真的有執行期消費端**（⛔ 不像 `lobby-layout` 那一份）：
 * `apps/client/src/ui/platform/lobbyRally.ts` 的 `activeLobbyRally()` 讀
 * `Configs.tryGet(LOBBY_RALLY_DOC_ID)`，和 `CameraRig` 讀 `config.camera@1`
 * 同一條路。⇒ 後台存檔 → 下一次載入內容就生效，⛔ 不必重建映像。
 *
 * ⚠️ 伺服器那一端**刻意只有柵欄，沒有政策**：`waitSeconds` 由客戶端送上去，
 * `internal/room/rally.go` 只把它夾進 [1, 120] 的傳輸界。理由和
 * `room.MatchSettings` 一模一樣 —— 界線抄第二份，drift 的那一份會安靜地拒絕
 * 主揪有權設定的值，而且看起來完全正常。
 */
/**
 * ⭐【`config.admin-friend@1`】—— **管理員預設好友**（GH#499）。
 *
 * owner 2026-08-21 逐字（三則）：
 *
 * > 「**所有人預設都會加管理員帳號為好友**」
 * > 「`adminAccountId` => **yes, 如果只有一個就預設那一個**」
 * > 「**管理員是強制雙向 不必請求 每個人創號自動預設有管理員好友**」
 *
 * ⭐ **強制雙向,⛔ 不是送出好友請求。** 平台走
 * `friend.Service.ForceFriend`（`apps/platform/internal/friend/friend.go`）
 * 在同一個兩邊一起寫的交易裡寫入雙方的 `friends` 邊,⛔ 不走 `Request()` ——
 * 那一條是「送出請求等對方接受」,對既有的 198 個帳號跑一次就會變成 **198 個
 * 沒有人會按的待處理請求**:功能上線、看起來做完了、而實際效果是零。
 *
 * ⚠️ **這一份的消費端是 Go 平台,⛔ 不是客戶端**（和這個檔案裡絕大多數文件不同）。
 * `friend.LoadAdminPolicy(CONTENT_DIR)` 在開機時讀出貨值,而後台存檔寫的是
 * **耐久覆蓋層**（`data/content-overlay/overlay.json`),所以 `AutoAdmin.Policy()`
 * **每一次決策都重讀一次覆蓋層**再疊上去 —— 少了那一步,這一頁就會變成 task #241
 * 那個形狀:操作者存了值、頁面顯示「✓ 已寫入」、重整還讀得回自己填的數字,而平台
 * 一輩子用出貨值（家用主機的 `content/` 是唯讀 bind-mount,連重啟都救不回來）。
 */
export const ADMIN_FRIEND_DOC_ID = "admin-friend";

export const zConfigAdminFriendDoc = z
  .object({
    id: z.literal(ADMIN_FRIEND_DOC_ID),
    schema: z.literal("config.admin-friend@1"),
    note: z.string().optional(),
    /**
     * 總開關。⭐ 預設 **on** —— owner 的原話是「所有人**預設**都會加管理員帳號
     * 為好友」,而第〇·六守則說「優先權大的更新後都是預設啟動」:開關存在是為了
     * **回頭**,不是為了觀望。
     *
     * ⛔ 關掉不會拆掉任何已經建立的好友關係（那是**移除**動作,而這一格是
     * 「以後還要不要繼續加」）。要拆得由玩家自己在好友名單移除。
     */
    enabled: z.boolean(),
    /**
     * ⭐ 大家會被加成好友的那一個管理員帳號 id。
     *
     * ⭐ **出貨值是字串 `"auto"`,⛔ 不是留空 —— 而 `"auto"` ≠ 關掉。**
     * owner 2026-08-21:「`adminAccountId` => **yes, 如果只有一個就預設那一個**」
     * ⇒ `"auto"` 是一條**會執行的規則**:
     *
     * ⚠️ **為什麼是 `"auto"` 而不是空字串**:後台通用表單的文字格拒絕空值
     * （`parseFieldInput`:「不可以是空的」）⇒ 出貨成 `""` 的話,操作者填了一次 id
     * 之後**再也回不到自動**,而那是一個只轉得動一邊的旋鈕（第一·五守則的形狀）。
     * 平台仍然把 `""` 當成 `"auto"` 的同義字（手改過的檔案讀得回來）,⛔ 但不出貨它。
     *
     * | 系統裡的 `RoleAdmin` 帳號數 | `"auto"` 時的行為 |
     * |---:|---|
     * | 1 | ⭐ 就是那一個（⛔ 不必填,也⛔ 不必每次換人時來改這一格） |
     * | 0 | 什麼都不做（新站的第一個帳號就是這個情況,開機回填會補上） |
     * | ≥2 | 什麼都不做,只記一行 log ——⛔ 平台**不替 owner 挑人**（見下面的隱私後果） |
     *
     * 填了 id 的時候是 **fail closed**:那個帳號讀不到、或它其實不是管理員,
     * 平台**不會**退回「自動挑一個」,而是什麼都不做並吼一聲 —— 一個打錯的 id
     * 靜靜地把全站好友加到別人身上,比不加更糟。
     *
     * ⚠️⚠️ **隱私後果（這一格真正的代價,⛔ 不是「多一個好友」）**：
     * 被指名的那個帳號會出現在**每一個玩家**的好友名單上,而好友名單會即時顯示
     * **在線狀態**（`online` / `in-lobby` / `in-match`）。⇒ 這一格等於
     * **讓這個帳號看得到全站每一個人現在在不在線上、正在大廳還是正在打**。
     * 它也讓那個帳號可以被每一個人看到同樣的資訊。⛔ 不要填一個不是站方的人。
     */
    adminAccountId: z.string().min(1).max(128),
    /**
     * ⭐ 開機時把**既有帳號**補一次。
     *
     * ⭐ 這是 owner 那句「**所有人**」的另一半:只接新帳號 = 今天以後註冊的人才有,
     * 而站上已經有 198 個帳號 —— 也就是**絕大多數人沒有**。
     *
     * 冪等:已經是好友的帳號一次寫入都不會發生,所以它可以每次開機跑。
     * 關掉它 = 只有新帳號會被接上;既有帳號改由後台 Quick Approval 的「加入」區
     * 手動觸發（`POST /api/v1/friends/admin-backfill`,只有管理員按得動）。
     */
    backfillExisting: z.boolean(),
    /**
     * 玩家**封鎖**了管理員（或反之）時,還要不要強制把好友加回去。
     *
     * ⭐ 出貨 **off**,這是 owner 沒有裁決到的那一格:「強制雙向」講的是
     * 「⛔ 不必請求」,而封鎖是一個玩家**明確按下去**的動作。打開它等於
     * 「封鎖對管理員無效」,那是一個站方應該自己決定、而不是被預設值決定的事。
     *
     * 打開時會連封鎖一起清掉（否則會留下一個「是好友、但也在封鎖名單上」的狀態,
     * 那是兩個互相矛盾的真相）。
     */
    overrideBlocked: z.boolean(),
  })
  .strict();

export type ConfigAdminFriendDoc = z.infer<typeof zConfigAdminFriendDoc>;

/** 去掉 id/schema/note 的殼之後,平台真正讀的那一份。 */
export type AdminFriendPolicyDoc = Omit<ConfigAdminFriendDoc, "id" | "schema" | "note">;

/**
 * 出貨預設（＝內容載不到時的保險絲）。
 *
 * ⚠️ 它必須和 `friend.DefaultAdminPolicy()`（Go）一字不差 —— 那一份才是平台在
 * 內容讀不到的時候真的會用的值。⭐ 兩邊都選 **on**:一份讀不到的內容文件如果
 * 靜靜地把功能關掉,「內容全毀」就會長得跟「owner 昨天關掉了這個功能」一模一樣,
 * 而兩個都不會有人看見（2026-08-01 骨架事故的形狀）。
 */
export const DEFAULT_ADMIN_FRIEND_POLICY: AdminFriendPolicyDoc = {
  enabled: true,
  adminAccountId: "auto",
  backfillExisting: true,
  overrideBlocked: false,
};

/** 文件 → 政策。缺席／壞掉一律回退到出貨預設。 */
export function resolveAdminFriend(
  doc: ConfigAdminFriendDoc | null | undefined,
): AdminFriendPolicyDoc {
  if (!doc) return DEFAULT_ADMIN_FRIEND_POLICY;
  return {
    enabled: doc.enabled,
    adminAccountId: doc.adminAccountId,
    backfillExisting: doc.backfillExisting,
    overrideBlocked: doc.overrideBlocked,
  };
}

File: apps/platform/internal/friend/adminfriend.go
package friend

// adminfriend.go — 管理員預設好友 (GH#499).
//
// ════════════════════════════════════════════════════════════════════════════
// owner 2026-08-21 逐字
// ════════════════════════════════════════════════════════════════════════════
//
//	「**所有人預設都會加管理員帳號為好友**」
//	「`adminAccountId` => **yes, 如果只有一個就預設那一個**」
//	「**管理員是強制雙向 不必請求 每個人創號自動預設有管理員好友**」
//
// ════════════════════════════════════════════════════════════════════════════
// 三件事,以及每一件為什麼不是它看起來的那個簡單做法
// ════════════════════════════════════════════════════════════════════════════
//
// ① ⛔ NOT `Service.Request()`. Request writes a PENDING edge the other side has
//    to accept. Running it for the existing accounts would have produced 198
//    「等待接受」rows nobody will ever press — a feature that ships, looks done,
//    and is worth exactly zero. `ForceFriend` writes both Friends edges in one
//    two-sided transaction instead (friend.go).
//
// ② ⛔ THE HOOK IS NOT IN auth.Register. It is `account.Repo.Create`'s
//    post-create seam, which runs only after the account FILE has landed. The
//    thing that must be impossible is「帳號沒建成、好友關係卻留下來了」, and any
//    seam placed earlier can produce exactly that. See account.PostCreateHook.
//
// ③ ⭐ NEW ACCOUNTS ALONE WOULD BE A NO-OP FOR ALMOST EVERYBODY. 198 accounts
//    already exist and none of them will ever be created again, so the backfill
//    (`Backfill`) is not a migration nicety — without it owner's「所有人」means
//    「今天以後註冊的人」. It runs at boot (Server.Boot, after the bootstrap-admin
//    grant so the admin it needs already carries the role) and is also exposed
//    as an admin-only route for the Quick Approval「加入」區 (handlers.go).
//    Idempotent by construction: ForceFriend reports changed=false when the two
//    are already friends, so boot number two writes nothing.
//
// ════════════════════════════════════════════════════════════════════════════
// ⚠️ WHY THE POLICY IS READ THROUGH THE OVERLAY AND NOT JUST content/
// ════════════════════════════════════════════════════════════════════════════
// The console saves a config doc through `PUT /content-overlay/docs/config/...`
// into data/content-overlay/overlay.json, while the shipped values live in
// content/config/admin-friend.json — and on the family host content/ is a
// READ-ONLY bind mount. A reader that only read content/ would give this page
// the exact shape task #241 documents: the operator saves, the page answers
// 「✓ 已寫入」, the page even re-renders the saved value, and the platform goes
// on using the shipped one forever. So shipped is the base and the overlay is
// layered on top HERE, on every decision, exactly like wallet/economy.go.

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"github.com/ggd/platform/internal/account"
	"github.com/ggd/platform/internal/data/jsonstore"
)

// Identifiers of the shipped policy document.
const (
	// AdminFriendDocID is the content doc id (content/config/admin-friend.json).
	AdminFriendDocID = "admin-friend"
	// AdminFriendSchema is its schema tag, mirroring zConfigAdminFriendDoc in
	// packages/shared/src/content/schema/config.ts.
	AdminFriendSchema = "config.admin-friend@1"
)

// Durable content-overlay identifiers, mirroring internal/contentoverlay's
// Collection / DocID / key(collection, id). COPIED, not imported: contentoverlay
// imports internal/admin, admin imports this layer's neighbours, and the edge
// would be an import cycle — the same copy wallet/economy.go keeps, for the same
// reason, with the same duty to be pinned by a test rather than trusted.
const (
	overlayCollection = "content-overlay"
	overlayDocID      = "overlay"
	// OverlayAdminFriendKey is what the console's putOverlayDoc("config",
	// "admin-friend") writes.
	OverlayAdminFriendKey = "config/" + AdminFriendDocID
)

// AdminPolicy is 管理員預設好友 as the platform actually reads it.
type AdminPolicy struct {
	// Enabled is the master switch. Off = no new-account link and no backfill.
	Enabled bool
	// AdminAccountID names the admin everybody is befriended with.
	//
	// ⭐ AdminAccountAuto (or empty) means「如果只有一個就預設那一個」— resolve it
	// from the roster. ⚠️ The shipped value is the WORD, not the empty string:
	// the console's text input refuses an empty value ("不可以是空的"), so an
	// operator who typed an id in could never get back to automatic. A knob you
	// can only turn one way is the shape of 第一·五守則.
	AdminAccountID string
	// BackfillExisting runs the existing-accounts pass at boot.
	BackfillExisting bool
	// OverrideBlocked forces the link through an explicit block. Shipped OFF —
	// see ForceFriend.
	OverrideBlocked bool
}

// DefaultAdminPolicy is the fuse: what the platform does when the document
// cannot be read at all. It is ON, because owner's sentence is「所有人預設都會加
// 管理員帳號為好友」and a content-read failure that silently disables the feature
// looks identical to owner having turned it off — neither of which anybody sees.
func DefaultAdminPolicy() AdminPolicy {
	return AdminPolicy{Enabled: true, AdminAccountID: AdminAccountAuto, BackfillExisting: true, OverrideBlocked: false}
}

// AdminAccountAuto is the AdminAccountID value that means「照 owner 的規則自己
// 判斷」. The empty string is accepted as a synonym (a hand-edited file, or the
// pre-console shape), but ⛔ never SHIPPED as one — see AdminPolicy.AdminAccountID.
const AdminAccountAuto = "auto"

// isAuto reports whether an adminAccountId asks for roster resolution.
func isAuto(id string) bool {
	t := strings.TrimSpace(id)
	return t == "" || strings.EqualFold(t, AdminAccountAuto)
}

// adminPolicyDoc is the wire shape. Every knob is a POINTER so a document that
// omits one keeps the shipped value instead of landing as Go's zero (which for
// three of the four booleans means "off" — a silent feature kill).
type adminPolicyDoc struct {
	Schema           string  `json:"schema"`
	Enabled          *bool   `json:"enabled"`
	AdminAccountID   *string `json:"adminAccountId"`
	BackfillExisting *bool   `json:"backfillExisting"`
	OverrideBlocked  *bool   `json:"overrideBlocked"`
}

func (p *AdminPolicy) apply(d adminPolicyDoc) {
	if d.Enabled != nil {
		p.Enabled = *d.Enabled
	}
	if d.AdminAccountID != nil {
		p.AdminAccountID = strings.TrimSpace(*d.AdminAccountID)
	}
	if d.BackfillExisting != nil {
		p.BackfillExisting = *d.BackfillExisting
	}
	if d.OverrideBlocked != nil {
		p.OverrideBlocked = *d.OverrideBlocked
	}
}

// LoadAdminPolicy reads the SHIPPED policy from content/config/admin-friend.json.
//
// ⚠️ Fail-open is deliberate (a missing content tree must not stop the platform
// booting) and therefore LOUD: both failure paths name the path they tried and
// the values they fell back to, because「內容讀不到」and「owner 關掉了這個功能」
// are otherwise the same silence.
func LoadAdminPolicy(contentDir string) AdminPolicy {
	p := DefaultAdminPolicy()
	if strings.TrimSpace(contentDir) == "" {
		return p
	}
	path := filepath.Join(contentDir, "config", AdminFriendDocID+".json")
	raw, err := os.ReadFile(path) //nolint:gosec // operator-supplied content dir
	if err != nil {
		slog.Warn("friend: 管理員預設好友 shipped policy is unreadable — using the built-in defaults",
			"path", path, "err", err, "enabled", p.Enabled, "backfillExisting", p.BackfillExisting)
		return p
	}
	var d adminPolicyDoc
	if err := json.Unmarshal(raw, &d); err != nil {
		slog.Error("friend: 管理員預設好友 shipped policy is not readable JSON — using the built-in defaults",
			"path", path, "err", err)
		return p
	}
	if d.Schema != AdminFriendSchema {
		slog.Error("friend: 管理員預設好友 shipped policy carries the wrong schema tag — using the built-in defaults",
			"path", path, "schema", d.Schema, "want", AdminFriendSchema)
		return p
	}
	p.apply(d)
	return p
}

// AutoAdmin is the wiring struct: it owns the shipped policy and the two things
// the feature needs (the social graph and the account roster).
type AutoAdmin struct {
	svc      *Service
	accounts *account.Repo
	shipped  AdminPolicy
}

// EnableAdminAutoFriend installs 管理員預設好友 on this service and returns the
// hook the account repository should be given. Composition root only.
//
// ⚠️ It stores itself on the Service so the admin-only backfill route can reach
// it without a fourth constructor argument threaded through three call sites of
// NewHandlers.
func (s *Service) EnableAdminAutoFriend(accounts *account.Repo, shipped AdminPolicy) *AutoAdmin {
	a := &AutoAdmin{svc: s, accounts: accounts, shipped: shipped}
	s.autoAdmin = a
	return a
}

// Policy is the live policy: shipped values with the operator's durable overlay
// layered on top. Re-read on every decision so a console save lands without a
// restart (see the file header).
func (a *AutoAdmin) Policy() AdminPolicy {
	p := a.shipped
	if a.svc == nil || a.svc.store == nil {
		return p
	}
	var f struct {
		Docs    map[string]json.RawMessage `json:"docs"`
		Deleted map[string]bool            `json:"deleted"`
	}
	if err := a.svc.store.Get(overlayCollection, overlayDocID, &f); err != nil {
		if !errors.Is(err, jsonstore.ErrNotFound) {
			slog.Warn("friend: could not read the content overlay — using the shipped 管理員預設好友 policy", "err", err)
		}
		return p
	}
	if f.Deleted[OverlayAdminFriendKey] {
		return p // operator explicitly reverted to shipped
	}
	raw, ok := f.Docs[OverlayAdminFriendKey]
	if !ok {
		return p
	}
	var d adminPolicyDoc
	if err := json.Unmarshal(raw, &d); err != nil {
		slog.Warn("friend: 管理員預設好友 override is not readable JSON — using the shipped policy",
			"key", OverlayAdminFriendKey, "err", err)
		return p
	}
	if d.Schema != AdminFriendSchema {
		slog.Warn("friend: 管理員預設好友 override carries the wrong schema tag — using the shipped policy",
			"key", OverlayAdminFriendKey, "schema", d.Schema, "want", AdminFriendSchema)
		return p
	}
	p.apply(d)
	return p
}

// ErrNoAdminResolved is returned when the policy names nobody and the roster
// cannot answer「只有一個就預設那一個」on its own.
var ErrNoAdminResolved = errors.New("friend: no administrator account could be resolved")

// ResolveAdminID answers 「誰是那個管理員」.
//
// ⭐ owner 2026-08-21:「`adminAccountId` => **yes, 如果只有一個就預設那一個**」.
// So "auto" is not「未設定,什麼都別做」— it is a live rule:
//
//	named in the policy → that account, but ONLY if it exists AND carries
//	                      RoleAdmin (fail CLOSED: a stale id must not quietly
//	                      befriend everybody with a deleted account, and a
//	                      non-admin id must not become an admin by being typed
//	                      into this box)
//	exactly one admin   → that one
//	zero / more than one → ErrNoAdminResolved, loudly, naming the count. More
//	                      than one is the case owner reserved for himself
//	                      (「多於一個時才需要他挑」) and picking one here would be
//	                      choosing on his behalf which human sees everybody's
//	                      online status.
func (a *AutoAdmin) ResolveAdminID(ctx context.Context, p AdminPolicy) (string, error) {
	if id := strings.TrimSpace(p.AdminAccountID); !isAuto(id) {
		acct, err := a.accounts.GetByID(ctx, id)
		if err != nil {
			return "", fmt.Errorf("friend: 管理員預設好友 names account %q, which cannot be read: %w", id, err)
		}
		if !acct.HasRole(account.RoleAdmin) {
			return "", fmt.Errorf("friend: 管理員預設好友 names account %q, which is not an administrator: %w", id, ErrNoAdminResolved)
		}
		return acct.ID, nil
	}
	admins, err := a.accounts.Admins(ctx)
	if err != nil {
		return "", err
	}
	if len(admins) == 1 {
		return admins[0], nil
	}
	return "", fmt.Errorf("friend: %d administrator accounts exist, so 管理員預設好友 needs adminAccountId set in the console: %w",
		len(admins), ErrNoAdminResolved)
}

// AfterAccountCreated is account.PostCreateHook: every account that durably
// lands is befriended with the administrator, both ways, immediately.
//
// Best-effort AND loud (see account.PostCreateHook): it can never fail the
// registration, so the only way anybody learns it did not happen is this log —
// plus the backfill, which will pick the account up on the next boot.
func (a *AutoAdmin) AfterAccountCreated(ctx context.Context, accountID string) {
	p := a.Policy()
	if !p.Enabled {
		return
	}
	adminID, err := a.ResolveAdminID(ctx, p)
	if err != nil {
		// ⚠️ Info, not Error: on a fresh deploy the FIRST account is created
		// before any admin exists, which is the ordinary path, not a fault.
		// Boot's backfill closes the gap.
		slog.Info("friend: 管理員預設好友 skipped for a new account — no administrator to link to yet",
			"accountId", accountID, "err", err)
		return
	}
	if adminID == accountID {
		return // the administrator's own account; withBoth would reject it anyway
	}
	if _, err := a.svc.ForceFriend(ctx, adminID, accountID, p.OverrideBlocked); err != nil {
		slog.Error("friend: could not link a new account to the administrator; it was created WITHOUT the default friend",
			"accountId", accountID, "adminId", adminID, "err", err)
		return
	}
	slog.Info("friend: 管理員預設好友 linked a new account", "accountId", accountID, "adminId", adminID)
}

// BackfillResult reports one existing-accounts pass.
type BackfillResult struct {
	AdminID string `json:"adminId"`
	Scanned int    `json:"scanned"`
	Linked  int    `json:"linked"`
	Failed  int    `json:"failed"`
}

// Backfill links every EXISTING account to the administrator (owner's ⭐: 「所有
// 人」, not「今天以後註冊的人」). Idempotent — an account that is already linked
// costs one read and zero writes.
//
// A per-account failure is counted and logged, never fatal: one unreadable file
// must not stop the other 197 from being linked.
func (a *AutoAdmin) Backfill(ctx context.Context) (BackfillResult, error) {
	p := a.Policy()
	if !p.Enabled {
		return BackfillResult{}, nil
	}
	adminID, err := a.ResolveAdminID(ctx, p)
	if err != nil {
		return BackfillResult{}, err
	}
	ids, err := a.accounts.List(ctx)
	if err != nil {
		return BackfillResult{}, err
	}
	res := BackfillResult{AdminID: adminID}
	for _, id := range ids {
		if id == adminID {
			continue
		}
		res.Scanned++
		changed, err := a.svc.ForceFriend(ctx, adminID, id, p.OverrideBlocked)
		if err != nil {
			res.Failed++
			slog.Error("friend: 管理員預設好友 backfill could not link one account", "accountId", id, "err", err)
			continue
		}
		if changed {
			res.Linked++
		}
	}
	return res, nil
}

// BackfillAdminFriendsInBackground runs Backfill off the boot path.
//
// ⚠️ It is a goroutine because it is O(accounts) file reads and boot must not
// wait on it; it is NOT fire-and-forget, because a backfill that silently did
// nothing is indistinguishable from one that linked everybody. Both outcomes
// are logged with counts.
func (s *Service) BackfillAdminFriendsInBackground(ctx context.Context) {
	a := s.autoAdmin
	if a == nil {
		return
	}
	if p := a.Policy(); !p.Enabled || !p.BackfillExisting {
		return
	}
	s.bg.Add(1)
	go func() {
		defer s.bg.Done()
		res, err := a.Backfill(ctx)
		if err != nil {
			slog.Warn("friend: 管理員預設好友 backfill did not run", "err", err)
			return
		}
		slog.Info("friend: 管理員預設好友 backfill finished",
			"adminId", res.AdminID, "scanned", res.Scanned, "linked", res.Linked, "failed", res.Failed)
	}()
}
